Added JsonResourceItems, Updated Responses, added ResourceTransformer.

davidxhill opened the pull request.

rmarganti reviewed (changes requested):
This PR was very difficult to parse, because there was no info on what was changing and why. No PR description, no updated readme, very few code comments. The only thing that really had my back was the tests.

Definitely update the README with some info on how these new abstractions (like `JsonApiResource`) are expected to be used in practice. And some details on how `fields` and `excludes` should be used.

davidxhill commented:
#### Added `excluded` as a request param option. (not officially supported by JsonAPI)
Attributes may be excluded from JsonApiResource's response. A decent use case for this would be removing the config/ui/tracking from Campaigns when it's not needed. This would allow us to remove some redundant routes, providing the FE with more flexibility.

Example:
`/api/export/campaigns?excludes[campaigns]=config,ui`

#### Added `fields` as a request param option. (https://jsonapi.org/format/#fetching-sparse-fieldsets)

> An empty value indicates that no fields should be returned.
> ...
> If a client requests a restricted set of fields for a given resource type, an endpoint MUST NOT include additional fields in resource objects of that type in its response.

Example:
`/api/permissionGroups?fields[permissionGroups]=name&include=permissions&fields[permissions]=`
```
{
  "data": [
    {
      "type": "permissionGroups",
      "id": "adops",
      "attributes": { "name": "adops" },
      "relationships": {
        "permissions": {
          "data": [{ "type": "permissions", "id": "campaigns.tracking" }]
        }
      }
    },
  ],
  "included": [
    { "type": "permissions", "id": "campaigns.tracking", "attributes": {} }
  ]
}
```

#### Added `JsonApiResource`
Forces entities to have the following methods: 
- `getResourceKey`: We can now reference this instead of hard coding this value very time. I added a new response method to be used in place of `withItem`, removing the need to pass resourceKey.
- `id`: JsonAPI allows requires an id.
- `jsonSerialize`: Used in the new ResourceTransformer.

#### To allow for easier and consistent exporting, I created a new ResourceTransformer that the following features:
- auto `excludes`/`fields` handling.
- added `transform` method to remove redundant code. 
- added `defaultFields` variable: since I expanded the `jsonSerialize` to include all attributes on entities...this will allow for easy defaults on transformers and still allow RequestParams to modify whats included. (see `CampaignBasicTransformer` for example)

davidxhill commented:
100% need to update the README good call.

rmarganti commented on src/Fractal/FieldsModifier.php:
```diff
@@ -0,0 +1,22 @@
+use Giadc\JsonApiRequest\Requests\RequestParams;
+use Giadc\JsonApiResponse\Interfaces\AbstractDataModifier;
+
+class FieldsModifier implements AbstractDataModifier
+{
+    public function __invoke(array $data, string $resourceKey, RequestParams $requestParams): array
```
To prevent our client and API from getting out of sync, it would be great if a validation error was thrown if the client requests/excludes fields that do not exist.

rmarganti commented on src/Fractal/ResourceTransformer.php:
```diff
@@ -0,0 +1,94 @@
+{
+    protected RequestParams $requestParams;
+
+    protected array $defaultFields = [];
```
Not fully-formed impression:

This can get very repetitive, with `$defaultFields` just being a duplication of most of the keys returned from a `jsonSerialize()` function. Do you think it could possibly useful if a `ResourceTransformer` had its own `fields` and `excludes` functionality? It may be less verbose to simply list the fields you _don't_ want included.

Counterpoint to my own point: being verbose helps avoid unexpected data leaks.

davidxhill replied:
I've been split on this myself...So I just defaulted to being more verbose for that reason. I am not sold on either option..and will add it if you want it.

rmarganti replied:
I think I'm cool with the verbosity. It's still better than our previous `transform()` functions that needed to manually pick out data.

rmarganti commented on tests/unit/Fractal/ResourceTransformerTest.php:
```diff
@@ -0,0 +1,81 @@
+
+    public function test_it_throws_exception_when_providing_invalid_fields(): void
+    {
+        try {
+            $request = Request::create(
+                'http://test.com/articles'
+                    . '?fields[tests]=name,ketchup'
+            );
+
+            $transformer = new TestTransformer($request);
+            $testEntity = new TestEntity(1, 'name');
+
+            $transformer->transform($testEntity);
+        } catch (\Exception $e) {
+            $this->assertInstanceOf(InvalidAttributesRequested::class, $e);
+            $this->assertEquals('Invalid Attribute Requested: ketchup. Valid keys are id, name, title.', $e->getMessage());
+        }
```
This test won't fail if the Transformer doesn't throw an Exception like we expect. Use PHPUnit's exception assertions.

```suggestion
        $this->expectException(InvalidAttributesRequested::class);
        $this->expectExceptionMessage('Invalid Attribute Requested: ketchup. Valid keys are id, name, title.');

        $request = Request::create(
            'http://test.com/articles'
                . '?fields[tests]=name,ketchup'
        );

        $transformer = new TestTransformer($request);
        $testEntity = new TestEntity(1, 'name');

        $transformer->transform($testEntity);
```
